delira/training/experiment.py

```diff
@@ -1,388 +1,582 @@
-from .parameters import Parameters
-from ..models import AbstractNetwork
-from .abstract_trainer import AbstractNetworkTrainer
-from trixi.experiment import Experiment as TrixiExperiment
-import os
-import logging
 import typing
-import numpy as np
-
+import logging
 import pickle
-from abc import abstractmethod
-from sklearn.model_selection import KFold, StratifiedKFold, StratifiedShuffleSplit
-from sklearn.model_selection import train_test_split
+import os
 from datetime import datetime
-from inspect import signature
 from functools import partial
 
+import numpy as np
+from sklearn.model_selection import KFold, StratifiedKFold, \
+    StratifiedShuffleSplit, ShuffleSplit
+from trixi.experiment import Experiment as TrixiExperiment
+
 from delira import get_backends
 
+from ..data_loading import BaseDataManager
+from ..models import AbstractNetwork
+
+from .parameters import Parameters
+from .base_trainer import BaseNetworkTrainer
+from .predictor import Predictor
+
 logger = logging.getLogger(__name__)
 
 
-class AbstractExperiment(TrixiExperiment):
+class BaseExperiment(TrixiExperiment):
     """
-    Abstract Class Representing a single Experiment (must be subclassed for
-    each Backend)
+    Baseclass for Experiments.
+
+    Implements:
+
+    * Setup-Behavior for Models, Trainers and Predictors (depending on train
+        and test case)
 
-    See Also
-    --------
-    :class:`PyTorchExperiment`
+    * The K-Fold logic (including stratified and random splitting)
+
+    * Argument Handling
 
     """
-    @abstractmethod
-    def __init__(self, n_epochs, *args, **kwargs):
+
+    def __init__(self,
+                 params: Parameters,
+                 model_cls: AbstractNetwork,
+                 n_epochs=None,
+                 name=None,
+                 save_path=None,
+                 key_mapping=None,
+                 val_score_key=None,
+                 optim_builder=None,
+                 checkpoint_freq=1,
+                 trainer_cls=BaseNetworkTrainer,
+                 **kwargs):
         """
 
         Parameters
         ----------
-        n_epochs : int
-            number of epochs to train
-        *args :
-            positional arguments
+        params : :class:`Parameters`
+            the training parameters
+        model_cls : Subclass of :class:`AbstractNetwork`
+            the class implementing the model to train
+        n_epochs : int or None
+            the number of epochs to train, if None: can be specified later
+            during actual training
+        name : str or None
+            the Experiment's name
+        save_path : str or None
+            the path to save the results and checkpoints to.
+            if None: Current working directory will be used
+        key_mapping : dict
+            mapping between data_dict and model inputs (necessary for
+            prediction with :class:`Predictor`-API)
+        val_score_key : str or None
```

Comment:
I think  it's more intuitive if `val_score_key` (currently set in `__init__`) and `val_score_mode`(currently set in `run`) would be set in the same function (maybe in `run`). What do you think about this?

Reply:
I think I left it that way, because I think this is what it was like until now, but I agree